lxd/device/gpu_sriov.go

```diff
@@ -0,0 +1,297 @@
+	saveData["last_state.vf.parent"] = pciParentDev.SlotName
+
+	vfID, err := d.findFreeVirtualFunction(pciParentDev)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to find free virtual function")
+	}
+
+	vfPCIDev, err := d.setupSriovParent(vfID, saveData)
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.volatileSet(saveData)
+	if err != nil {
+		return nil, err
+	}
+
+	runConf.GPUDevice = append(runConf.GPUDevice, []deviceConfig.RunConfigItem{
+		{Key: "devName", Value: d.name},
+		{Key: "pciSlotName", Value: vfPCIDev.SlotName},
+	}...)
+
+	return &runConf, nil
+}
+
+// setupSriovParent configures a SR-IOV virtual function (VF) device on parent and stores original properties of
+// the physical device into voltatile for restoration on detach. Returns VF PCI device info.
+func (d *gpuSRIOV) setupSriovParent(vfID int, volatile map[string]string) (pci.Device, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	volatile["last_state.vf.id"] = fmt.Sprintf("%d", vfID)
+	volatile["last_state.created"] = "false" // Indicates don't delete device at stop time.
+
+	// Get VF device's PCI Slot Name so we can unbind and rebind it from the host.
+	vfPCIDev, err := d.getVFDevicePCISlot(volatile["last_state.vf.parent"], volatile["last_state.vf.id"])
+	if err != nil {
+		return vfPCIDev, err
+	}
+
+	// Unbind VF device from the host so that the settings will take effect when we rebind it.
+	err = pci.DeviceUnbind(vfPCIDev)
+	if err != nil {
+		return vfPCIDev, err
+	}
+
+	revert.Add(func() { pci.DeviceProbe(vfPCIDev) })
+
+	// Register VF device with vfio-pci driver so it can be passed to VM.
+	err = pci.DeviceDriverOverride(vfPCIDev, "vfio-pci")
+	if err != nil {
+		return vfPCIDev, err
+	}
+
+	// Record original driver used by VF device for restore.
+	volatile["last_state.pci.driver"] = vfPCIDev.Driver
+
+	revert.Success()
+
+	return vfPCIDev, nil
+}
+
+// getVFDevicePCISlot returns the PCI slot name for a PCI virtual function device.
+func (d *gpuSRIOV) getVFDevicePCISlot(parentPCIAddress string, vfID string) (pci.Device, error) {
+	ueventFile := fmt.Sprintf("/sys/bus/pci/%s/device/virtfn%s/uevent", parentPCIAddress, vfID)
+	pciDev, err := pci.ParseUeventFile(ueventFile)
+	if err != nil {
+		return pciDev, err
+	}
+
+	return pciDev, nil
+}
+
+func (d *gpuSRIOV) findFreeVirtualFunction(parentDev pci.Device) (int, error) {
+	// Get number of currently enabled VFs.
+	sriovNumVFs := fmt.Sprintf("/sys/bus/pci/devices/%s/sriov_numvfs", parentDev.SlotName)
+
+	sriovNumVfsBuf, err := ioutil.ReadFile(sriovNumVFs)
+	if err != nil {
+		return 0, err
+	}
+	sriovNumVfsStr := strings.TrimSpace(string(sriovNumVfsBuf))
+	sriovNum, err := strconv.Atoi(sriovNumVfsStr)
+	if err != nil {
+		return 0, err
+	}
+
+	vfID := -1
+
+	for i := 0; i < sriovNum; i++ {
+		pciDev, err := pci.ParseUeventFile(fmt.Sprintf("/sys/bus/pci/devices/%s/virtfn%d", parentDev.SlotName, i))
+		if err != nil {
+			return 0, err
+		}
+
+		if pciDev.Driver == "" {
```

Comment:
@monstermunchkin could you add a comment here explaining why this means the VF is free (presumably its because its not bound to any driver?). thanks

Reply:
@tomponline I'll add a comment